vertx-mail-client/src/main/java/io/vertx/ext/mail/impl/MailClientImpl.java

```diff
@@ -0,0 +1,105 @@
+package io.vertx.ext.mail.impl;
+
+import io.vertx.core.*;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.impl.LoggerFactory;
+import io.vertx.ext.mail.MailClient;
+import io.vertx.ext.mail.MailConfig;
+import io.vertx.ext.mail.MailMessage;
+
+/**
+ * MailService implementation for sending mails inside the local JVM
+ *
+ * @author <a href="http://oss.lehmann.cx/">Alexander Lehmann</a>
+ */
+public class MailClientImpl implements MailClient {
+
+  private static final Logger log = LoggerFactory.getLogger(MailClientImpl.class);
+
+  private final Vertx vertx;
+  private final SMTPConnectionPool connectionPool;
+  private volatile boolean closed = false;
+
+  public MailClientImpl(Vertx vertx, MailConfig config) {
+    this.vertx = vertx;
+    this.connectionPool = new SMTPConnectionPool(vertx, config);
+  }
+
+  @Override
+  public void close() {
+    connectionPool.close();
+    closed = true;
+  }
+
+  @Override
+  public MailClient sendMail(MailMessage message, Handler<AsyncResult<JsonObject>> resultHandler) {
+    Context context = vertx.getOrCreateContext();
+    if (!closed) {
+      if (validateHeaders(message, resultHandler, context)) {
+        connectionPool.getConnection(conn -> sendMessage(message, conn, resultHandler, context),
+                                     t ->  handleError(t, resultHandler, context));
+      }
+    } else {
+      handleError("mail service has been closed", resultHandler, context);
+    }
+    return this;
+  }
+
+  private void sendMessage(MailMessage email, SMTPConnection conn, Handler<AsyncResult<JsonObject>> resultHandler,
+                           Context context) {
+    new SMTPSendMail(conn, email, v -> {
+      conn.returnToPool();
+      /// FIXME Why return a JSON object? What's the point?
```

Comment:
when sending mails to multiple recipients it is possible that some addresses work and some don't, for this an array or json object can be returned with the list of successful addresses and the list of failed addresses. this is not implemented yet